### ISSUE TITLE
Preprocess any data

### ISSUE THREAD

**Issue by 14corman** [14corman/AIDescisionTree#3]
Not sure yet if any data needs to be processed from the 4 datasets, but if it does then we need to do it.

**Comment by 14corman** [comment]
@the5thbeatle do you have to create the ml file from the python script? I thought the python script was just going to do any preprocessing of the data and output a new csv file or something. I would not recommend you make the ml file in that way. You cannot open it in a text editor because of its size if you need to edit, or look at, it. You should be able to open and work with those files in ocaml if you treat each line by splitting it by a ",' and going over those values. You could even place -1's in for "?". Then calculate modes and go back through the lists to replace "?" with those modes. I think it may be useful to create a ml file to load in and do everything required to these data files. Then you can make a ml file for each dataset that uses the load in file, use the mode function I have to finalize then, then do classifications.

**Comment by the5thbeatle** [comment]
Here was my mindset for creating the ml file. I believe the PDF said that
we could hardcode the datasets for ease, so I didn’t think they would be
looking for an OCaml csv parser. On top of that, the difference between
creating the csv and the ml file is entirely trivial. The portion of code
that creates the map from feature names to the examples’ values can be
easily modified to create a csv instead; however, now that extra data
processing on top of the preprocessing isn’t as necessary. Finally, a large
file isn’t an issue in itself, especially if it’s data that we should
consider read-only; it doesn’t make much sense to directly edit the files
containing our dataset anyway. Plus, the file’s size is directly
proportional to the size of the combined csvs. Replacing commas with a
semicolon and a space is the only non-constant change in size; the file
opens just as well in Vim as the csvs do.

I will add to the code so that it produces a csv that highly resembles the
format OCaml will put it into. I will also enter in the -1 in place of ?,
but as I said before, I believe that further handling of missing values
should be handled by the tree builder or classifier. Assigning values at
the beginning of the process throws away the opportunity to assign values
when the examples have been sorted into more closely related subgroups.

My biggest worry is to begin work on the cross-validation code. I’ll make
the changes to the python code, and then I’ll use the hardcoded datasets to
work on validation, which has the non-trivial task of shuffling a variable
number of lists the same way due to the examples not being the discrete
units.

**Comment by 14corman** [comment]
Ok I did not think of building the ml file from the data to be hardcoding it, but that is a good idea.

 I mentioned to do the -1 so that the decision tree code can use the mode to replace the -1's. If you think doing that in python would be better/easier than you should do that then.

I was going to work on doing this in my metrics file I was going to work on today. I was going to have it split on a certain point for train and test set. Then incrementally make the training set large and test set smaller to get the graph of how it is going. I could do this differently using cross validation by giving an integer k that will split the sets into k splits.

If you are worried about shufling, here is some code to do just that:
``` ocaml
let shuffle d =
    let nd = List.map (fun c -> (Random.bits (), c)) d in
    let sond = List.sort compare nd in
    List.map snd sond
```

You can change this to this:
``` ocaml
let shuffle d rand =
    let nd = List.map2 (fun r c -> (r, c)) rand d in
    let sond = List.sort compare nd in
    List.map snd sond
```

You can create 1 random list of bits that way each list is sorted in the same way. You just pass this and the list of feature examples and it will return a random shuffle of that list.

If you want to start making that file then I can work on the metrics file that will create F1 and accuracy.

**Comment by 14corman** [comment]
I am not sure what all you have been able to do with the cross validation code, but I have something put into my metrics file that I am working on. I have the F1 score already in there. If you want to work on it, let me know and I can push what I have for you to see. I was thinking of making a list of maps for the folds and a list of lists for the y folds. Then, you can just do an iterate 2 to build and classify. When you do that, you need to keep track of the list of predictions and send them into the f1_score function I have that will calculate the F1 score across alll fold configurations at once and return the 1 score.

Right now, I have the code able to split up lists into lists of sizes we want the folds to be. For example, say we have a list ```[0; 1; 1; 0; 0; 0; 0; 1; 0; 1; 0; 1; 1; 1; 0; 0; 1; 1; 0; 1]```, and you want to do f-fold cross validation. Then it will make ```[[0; 1; 1; 0]; [0; 0; 0; 1]; [0; 1; 0; 1]; [1; 1; 0; 0]; [1; 1; 0; 1]]```. It also has the shuffle method built in so it shuffles each list the same way randomly.

Final partitioning would be for example (with this feature name being "x1"):

    {"x1": [0; 1; 1; 0; 0; 0; 0; 1; 0; 1; 0; 1; 1; 1; 0; 0; 1; 1; 0; 1]}

would turn into

    [{"x1": [0; 1; 1; 0]}; {"x1": [0; 0; 0; 1]}; {"x1": [0; 1; 0; 1]}; {"x1": [1; 1; 0; 0]}; {"x1": [1; 1; 0; 1]}]

**Comment by 14corman** [comment]
I think we are done with this.
